### DISCUSSION TITLE
Access non geometry layer for Plugin

### DISCUSSION THREAD

**Discussion by paul-carteron** [opengisch/QField#5877] [Q&A]
Hi,

I'm attempting to access a non-geometrical layer to enumerate all values within a specific field. While I can retrieve `QgsVectorLayer `instances using `mapCanvas.mapSettings.layers`, I've noted that, as specified in the [QField source code](https://github.com/opengisch/QField/blob/f3c1fb9b351bba39ed7771efb486e898a11432e5/src/core/qgsquick/qgsquickmapsettings.h#L114)
> non-spatial layers are automatically excluded from this list.

Additionally, I've observed that `dashBoard.activeLayer` provides access to the currently active layer. However, my requirement is to access any layer within the project, not just the active one.

Is there a method to access non-geometrical layers within QField's QML environment?

Below my qml to test layer acces :
```
import QtQuick

Item {
  id: plugin

  property var mapCanvas: iface.findItemByObjectName('mapCanvas')

  Component.onCompleted: {

    let layers = mapCanvas.mapSettings.layers;
    let concatenatedLayerNames = layers.map(layer => layer.name).join(", ");
    iface.logMessage(`Layer Names: ${concatenatedLayerNames}`);

    let firstLayer = layers[0];
    iface.logMessage(`Layer Name: ${firstLayer.name}`);
    iface.logMessage(`Fields Name: ${firstLayer.fields.names}`);

    let fieldName = 'PLTM_PARC';
    iface.logMessage(`Index of ${fieldName} is ${firstLayer.fields.indexOf(fieldName)}`);
    
  }
}
```

**Comment by nirvn** [comment]
@paul-carteron , have you tried this function: https://api.qgis.org/api/classQgsProject.html#a935efacacb942a15afe19728233e35c3 ? In the qml plugin environment you can access the current project via the qgisProject object. 

So provided you know your layer name you could do qgisProject.mapLayersByName("my layer").

**Reply by nirvn** [reply]
Pro tip: in the QGIS documentation, you can see properties accessible in QML within the properties section, and you can see which functions are exposed by spotting the Q_INVOKABLE tags.

**Comment by paul-carteron** [comment]
@nirvn It works perfectly, thank you!

I wasn’t aware that `qgisProject` is accessible in QField plugins. What I take away after watching your [presentation](https://drive.google.com/file/d/1s5SD-JpCq9QHoIbmOMEqDsTyGSsN6QrC/view), I that i can use in Qfield Plugin :
- [All Q_INVOKABLE iface functions](https://github.com/opengisch/QField/blob/master/src/core/appinterface.h) ;
- [Any QML components defined in qml/qgismobileapp.qml](https://github.com/opengisch/QField/blob/master/src/qml/qgismobileapp.qml), as long as they have an `objectName` property ;
-  [Utility classes from core/utils](https://github.com/opengisch/QField/tree/master/src/core/utils), offering reusable functions for common tasks ;
- [QField custom QML components from Theme](https://github.com/opengisch/QField/tree/master/src/qml/imports/Theme), which include various UI and styling elements ;
- [Qgs classes initialized in core/qgismobileapp.cpp](https://github.com/opengisch/QField/blob/f3c1fb9b351bba39ed7771efb486e898a11432e5/src/core/qgismobileapp.cpp#L365), specifically within the initDeclarative function ;
- (and also the pro tip you just gave me) 

You - obviously - already know all this, but I'm writing more with the goal of documenting and centralizing ressources. With that in mind, I couldn’t find where it’s explicitly documented that `qgisProject` exposes the QgsProject class. Do you know where I can find where this is defined?

**Reply by nirvn** [reply]
@paul-carteron , you'd see it exposed in the QML context here: https://github.com/opengisch/QField/blob/f3c1fb9b351bba39ed7771efb486e898a11432e5/src/core/qgismobileapp.cpp#L575

We'd most welcome documentation additions once you feel happy about your gathering of resources.

**Reply by paul-carteron** [reply]
Thanks, I'll keep that in mind for the documentation suggestion !
